python/mspasspy/db/database.py

```diff
@@ -570,6 +571,150 @@ def update_ensemble_metadata(self, ensemble_object, include_undefined=False, exc
+        :type mspass_object:  must be TimeSeries, Seismogram, TimeSeriesEnsemble, or SeismogramEnsemble.
+        
+        :param exclude_keys: list of attributes that should not normally be loaded.
+        Default attributes not normally need that are loaded from stationxml.  Ignored if include_undefined is set True.
+        
+        :param include_undefined:  when true all data in the matching document are loaded.
+
+        :param collection: requested collection metadata should be loaded
+        :type collection: str 
+
+        :exception:  any detected errors will cause a MsPASSError to be thrown
+        (colleagues:  this may be wrong sphynx syntax for defining an exception)
+        """
+        if collection not in ["source", "site", "channel"]:
+            raise MsPASSError("collection param should be one of source, site or channel", ErrorSeverity.Invalid)
+
+        if not mspass_object.live:
+            raise MsPASSError("only live mspass object can load metadata", ErrorSeverity.Invalid)
+
+        if not isinstance(mspass_object, (TimeSeries, Seismogram)):
+            raise MsPASSError("only TimeSeries and Seismogram are supported", ErrorSeverity.Invalid)
+
+        if collection == 'channel' and isinstance(mspass_object, (Seismogram, SeismogramEnsemble)):
+            raise MsPASSError("channel data can not be loaded into Seismogram", ErrorSeverity.Invalid)
+        
+        # 1. get the metadata schema based on the mspass object type
+        wf_collection = 'wf_TimeSeries' if isinstance(mspass_object, TimeSeries) else 'wf_Seismogram'
+        object_type = self.database_schema[wf_collection].data_type()
+        if object_type not in [TimeSeries, Seismogram]:
+            raise MsPASSError('only TimeSeries and Seismogram are supported, but {} is requested. Please check the data_type of {} collection.'.format(
+                object_type, wf_collection), 'Fatal')
+        wf_collection_metadata_schema = self.metadata_schema[object_type.__name__]
+
+        collection_id = collection + '_id'
+        # 2. get the source_id from the current mspass_object
+        if not mspass_object.is_defined(collection_id):
+            raise MsPASSError("no {} in the mspass object".format(collection_id), ErrorSeverity.Invalid)
+        object_doc_id = mspass_object[collection_id]
+
+        # 3. find the unique document associated with this source id in the source collection
+        object_doc = self[collection].find_one({'_id': object_doc_id})
+        if object_doc == None:
+            raise MsPASSError("no match found in {} collection for source_id = {}".format(collection, object_doc_id), ErrorSeverity.Invalid)
+        
+        # 4. use this document to update the mspass object
+        for k in wf_collection_metadata_schema.keys():
+            col = wf_collection_metadata_schema.collection(k)
+            if col == collection:
+                if include_undefined:
```

Comment:
I don't think `include_undefined` is functioning here. The loop is driven by `wf_collection_metadata_schema.keys()` which are all the defined keys. The undefined keys in `object_doc` will never be reached.

Reply:
I am a little bit confused about the scope of `include_undefined` and `exclude_keys` parameters. From my understanding, `include_undefined` should include the keys in the `object_doc`, while `exclude_keys` excludes the keys in the metadata. Correct me if I am wrong.